Add documentation for FileWriter V2 (#1)
soph-dec/documentation: filewriter_v2/README.md

# FileWriter V2

---
$\color{red}\text{\large{NOTE: FileWriter V2 is currently in a preview state and will change in the future.}}$

Feedback is very welcome!

---

FileWriter writes image data and metadata to [HDF5](https://www.hdfgroup.org/solutions/hdf5/) files. Currently, there are two different formats available. V1 is a legacy format that is loosely based on [version 1.4](https://github.com/nexusformat/definitions/releases/tag/NXmx-1.4) of the _NXmx application definition_ of the [NeXus] format. V2 follows the [_NXmx application definition_](https://manual.nexusformat.org/classes/applications/NXmx.html) of NeXus Version 2022.07, with the addition of multi-channel data support[^1].

> | Version | Configuration | NeXus/NXmx Version | Multi-channel data support |

Reviewer: Simplify "Multi-channel data support" to "Multi-channel data"?

Also use consistent capitalization in header rows. Pick one of:
* title-style: "NeXus/NXmx Version", "Multi-Channel Data Support", "JSON Type"
* sentence-style: "NeXus/NXmx version" + "Multi-channel data support", "JSON type"

NOTE: I used title-style capitalization for this in the stream V2 documentation.

Author: done in [c2d9663](https://github.com/soph-dec/documentation/pull/1/commits/c2d9663406afc29c55634d16dc0acb50f6f81ddc)

|----|----|----|----|

> | V1 | SIMPLON | loosely based on [NXmx 1.4](https://github.com/nexusformat/definitions/releases/tag/NXmx-1.4) | no |

Reviewer: I am wondering if this is the best wording. How "loosely" is V1 based NXmx 1.4? Are there many examples where we don't fulfill that standard? I know we provide extra information throught "dectrisSpecific", however, that may not be a problem for the under-specified NXmx 1.4.

Also it might be a good idea to include the year after "NXmx 1.4" to make it more obvious that its an old defunct standard. We could write something like "Legacy NXmx 1.4 (2016)"?

Author: I will take a closer look at version 1.4, change the wording if necessary and add the year.

Author: I checked now if all the fields without `minOccurs="0"` are present for fw v1 files. I think the only thing that is different is that the data is not in the detector group. So it is more than loosely based on it. I am not 100% sure if detector/data was required or not, since I can't build the docs without python2 (and don't want to install it).

Reviewer: You can't tell from the source code if it was required? If that's the only thing missing (and especially if it's not clear that detector/data is required), then maybe we should write "NeXus NXmx 1.4 (2016)"?

Author: Now I have found these:
https://github.com/nexusformat/definitions/tree/main/legacy_docs
In the two 2016 versions, NXmx is listed as version 1.4, it looks like detector/data was required then.
Another thing I found was that fw v1 writes "flatfield_correction_appplied", nxmx defines "flatfield_applied", but since it is optional, I don't see a real problem here.

I think all except detector/data is following NXmx 1.4.

Author: Should we reword this to "mostly following NXmx 1.4 (2016)", "very close to NXmx 1.4 (2016)", "almost NXmx 1.4 (2016)"? XD

Reviewer: I think I would avoid language like "mostly" or "very close". Wouldn't such language also apply to V2 since things like `source_name` are also technically required?

Author: Let's go with "NeXus NXmx 1.4 (2016)". I'll link to the (older) pdf in the legacy docs.

Author: I changed it in [560ea19](https://github.com/soph-dec/documentation/pull/1/commits/560ea1908a64c39f1b9d817679495d51ecd370ae).

Reviewer: Thanks. But why not link to the later document [nexus-v3.2-2016-11-22-32b130a.pdf](https://github.com/nexusformat/definitions/blob/main/legacy_docs/nexus-v3.2-2016-11-22-32b130a.pdf) instead (still NXmx 1.4).

Reviewer: Looking at the linked document, I just realized nxmx 1.4 was also part of NeXus v3.2. So that is another possibility (perhaps better?): "NeXus 3.2 NXmx (2016)". What do you think?

Author: Oh, I didn't see that. Yes, its more consistent and makes sense to write "NeXus 3.2 NXmx (2016)".
As for what document to link, there does not seem to be a difference in the NXmx definition, so I think both are fine. If you prefer the second one, I can change it. The tag "NXmx-1.4" is somewhere in between the two versions.

Author: Changed the link, the wording of the text above and the version name [here](https://github.com/soph-dec/documentation/pull/1/commits/05b1207e7a03334adbda2e409c0cfbdc6fee192b).

> | V2 | SIMPLON | [NeXus 2022.07 NXmx](https://github.com/nexusformat/definitions/blob/v2022.07/applications/NXmx.nxdl.xml) | yes |

Reviewer: Capitalize "No/Yes" in the table.

Author: don in [c2d9663](https://github.com/soph-dec/documentation/pull/1/commits/c2d9663406afc29c55634d16dc0acb50f6f81ddc)

[^1]: This is not part of the NeXus standard yet, a detailed PR will be opened when this data format has been approved by the community. The general idea of this representation has already been discussed and approved by the NIAC though, see this [issue](https://github.com/nexusformat/definitions/issues/940).

> Therefore, a FileWriter V2 file is structured into the following groups:

Reviewer: Remove "therefore". Simplify to: "FileWriter V2 files are structured into the the following (HDF5) groups:".

Also I think the tree feels a bit out of place here. I would move it down lower above "List of fields".

Author: Done in [da8da81](https://github.com/soph-dec/documentation/pull/1/commits/da8da8177a76e0a4123658ba94eaf674b647431b)

```
/
|
+-- entry
    |
    +-- data
    +-- instrument
    |   |
    |   +-- beam
    |   +-- detector
    |       |
    |       +-- threshold_1_channel*
    |       +-- threshold_2_channel*
    |       +-- difference_channel*
    |       +-- geometry
    |       +-- module
    |       +-- transformations
    |
    +-- sample
    |   |
    |   +-- transformations
    |   +-- beam
    |
    +-- source
```
\* this is not part of the NeXus standard yet, see [Multi-Channel Data](#multi-channel-data).

## FileWriter V2 Configuration

FileWriter V2 can be enabled and configured using the SIMPLON API.

> ### SIMPLON FileWriter Configuration Parameters

Reviewer: I don't think we need to write "FileWriter" here since it's obvious from context (it is a subsection of "FileWriter V2 Configuration"). I did the same thing in the stream V2 documentation. It could be simplified there too.

Author: I copied this from stream so that the documentations would match. I can change both of them.

Author: Done in [770c663](https://github.com/soph-dec/documentation/pull/1/commits/770c6633d39c11c03e5d11c9d9a7cd206809d626).

The configuration interface is accessible at `http://<ADDRESS_OF_DCU>/filewriter/api/1.8.0/config`.

| Parameter | JSON Type | Default | Description |
|-----------|------|---------|-------------|

> | `mode` | string | `disabled` | Enabled state of the FileWriter, either `enabled` or `disabled`. |

Reviewer: I would write `"disabled"` (include quotes in value text) since this is the valid JSON representation.

Author: Changed in [da0fe48](https://github.com/soph-dec/documentation/pull/1/commits/da0fe48b781e4efb7d196bab0f879f5f15b483ee).

| `compression_enabled` | bool | `true` | Optionally disable compression of image data. |

> | `image_nr_start` | uint | `1` | First image number on the axis `/entry/image_number`. |

Reviewer: "uint" is not a JSON type. There are two ways to resolve this:

1. The column specifies the JSON type and any further restrictions are specified in the description. In this case, you would specify "number" instead of "uint" and "boolean" instead of "bool". The set of JSON types is: "string", "number", "boolean", "null", "object", "array".
2. Specify SIMPLON types in the column instead (e.g. change the column header to "SIMPLON Type").

Author: Good point, I wrote the types the way they are written in the SIMPLON documentation. I think I prefer to go with the JSON types and add restrictions in the description section, see [da0fe48](https://github.com/soph-dec/documentation/pull/1/commits/da0fe48b781e4efb7d196bab0f879f5f15b483ee).

> | `name_pattern` | string | `series_$id` | Base name of the files. `$id` will be replaced by the series number, i.e. using the default, the generated files will have the names:<br> `series_<series_number>_master.h5`<br> `series_<series_number>_data_<file_number>.h5`|

Reviewer: `"series_$id"` (include quotes)

Author: Fixed in [da0fe48](https://github.com/soph-dec/documentation/pull/1/commits/da0fe48b781e4efb7d196bab0f879f5f15b483ee).

| Parameter | JSON Type | Default | Description |
|-----------|------|---------|-------------|
| `enabled` | boolean | `false` | Enabled state of the FileWriter. |
| `image_id_offset` | uint | `1` | First image number on the axis `/entry/image_number`. |
| `name_pattern` | string | `series_$id` | Base name of the files. `$id` will be replaced by the series number, i.e. using the default, the generated files will have the names:<br> `series_<series_number>_master.h5`<br> `series_<series_number>_data_<file_number>.h5`|
| `images_per_file` | uint | `1000` | Maximum number of images stored in each data file. If set to `0`, all images are stored directly in the master file and no data files are created. <br> If set to a value greater than zero, the images are stored in multiple data files, but are still accessible in the master file via the dataset "/entry/data/data", since this is a [Virtual Dataset](https://portal.hdfgroup.org/display/HDF5/Virtual+Dataset++-+VDS) in this case.|
| `v2_preview_enabled` | bool | `false` | Enabled state of the FileWriter V2 preview. |
-->

### Example

FileWriter V2 may be configured with cURL.

> Using SIMPLON:

Reviewer: This line seems extraneous given that we are in section "SIMPLON FileWriter Configuration Parameters". Also, SIMPLON is the only official interface we have.

Author: Ah, this is leftover from when I thought we would have a SIMPLON V2 example as well. I'll remove it.

Author: Fixed in [ac4aa16](https://github.com/soph-dec/documentation/pull/1/commits/ac4aa16eed467e1388a7efaa1bc9cf0a37abc8f6).

```sh
curl \
    -X PUT \
    -H "Content-Type: application/json" \
    -d '{"mode": {"value": "enabled"}, "v2_preview_enabled": {"value": true}}' \
    $ADDRESS_OF_DCU/filewriter/api/1.8.0/config
```

## Data Access

> The files are stored locally on the DCU and have to be transferred to the user computer, this can also be done via cURL using the SIMPLON API.

Reviewer: Maybe this can be clarified some. Like:

Files are stored locally (in memory) on the DCU and may be accessed over HTTP. Files are NOT stored persistently and all data will be lost on reboot.

## Example

Author: I changed it according to your suggestion ([e3f9f80](https://github.com/soph-dec/documentation/pull/1/commits/e3f9f807d7d95826c58c9c56679085adece5b55b)). I only changed the all-caps "NOT" to a bold "**not**".

Reviewer: Looks good. You could also consider making the whole second sentence bold instead?

Author: Either is fine with me, if you prefer the whole sentence to be bold I'll change it for you.

Reviewer: I just looked and at least on github you can barely see the bold with the dark theme. Can we make the whole sentence bold and red? Seems like an important detail that should be clear to the customer. I see this question come up quite often internally too.

Author: Colors are a bit tricky, I think I'd have to use the Latex workaround like in the beginning, which results in a different font:
$\color{red}\textsf{Files are not stored persistently and all data will be lost on reboot.}$
or 
$\color{red}\textbf{Files are not stored persistently and all data will be lost on reboot.}$

I could also use something like this:

> **Warning**
> Files are **not** stored persistently and all data will be lost on reboot.

Author: I went with the warning and also used that in the very beginning for the preview warning. See [f64a692](https://github.com/soph-dec/documentation/pull/1/commits/f64a6925f5a2c65710769a5fec730958f5cc3ffd).

```sh
# list available files
curl $ADDRESS_OF_DCU/filewriter/api/1.8.0/status/files

# file access
curl -O $ADDRESS_OF_DCU/data/<name_pattern>_master.h5
curl -O $ADDRESS_OF_DCU/data/<name_pattern>_data_<file_number>.h5
```

For further details on available commands and status parameters please refer to the [SIMPLON API documentation](https://media.dectris.com/210607-DECTRIS-SIMPLON-API-Manual_EIGER2-chip-based_detectros.pdf).

> ## Multi-channel Data

Reviewer: Either "Multi-Channel Data" or "Multi-channel data". If you want to use sentence-style capitalization for sections then at least do it consistently (e.g. "Data access" above). I used title-style capitalization for section names/table headers for stream V2.

Author: Done in [5e7e4cb](https://github.com/soph-dec/documentation/pull/1/commits/5e7e4cbdb401b1b6deb74a9db3a646bf74a7863b)

> One of the main new features of FileWriter V2 is the support of multi-channel data, i.e. each image consists of multiple channels, one per enabled threshold/difference mode. Therefore, the dataset `/entry/data/data` is four-dimensional, i.e. `[number_of_images, number_of_channels, width, height]`. The channels are identified via strings given in the [_AXISNAME_ field](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-axisname-field) `channel` of the _NXdata_ class[^2]. For every channel name listed in this field, there exists a subgroup of the [_NXdetector_]() group of type _NXdetector_channel_[^1] with the same name followed by the suffix `_channel`. This group contains channel-specific metadata, for example `threshold_energy`, `flatfield` or `pixel_mask`.

Reviewer: Swap "width" and "height" (the order is wrong). Also be careful not to abuse the verbatim style. The dimensions should not be written out verbatim (e.g. there is no name "number_of_channels"). Also, unless you are referring to the actual name of something (like with verbatim), I don't think underscores are appropriate.

Maybe we can start with something like (or at least here is some inspiration):

Acquired images may consist of multiple channels, one per enabled threshold/difference mode. FileWriter V2 introduces support for multi-channel image data. The dataset `/entry/data/data` is four-dimensional with shape [_number of images, number of channels, height, width_]. Channels are identified via...

Maybe instead of writing out the shape, the dimensions should just be named directly like: [_image number, channel index, y, x_]? Just a thought.

I also suggest rewording the "AXISNAME field `channel`" part. It seems to be quite confusing to read.

Author: Thanks, using the verbatim style only for existing names is a very good point. How about this?

Acquired images may consist of multiple channels, one per enabled threshold/difference mode. FileWriter V2 introduces support for multi-channel image data. The dataset `/entry/data/data` is four-dimensional with shape $[n, m, h, w]$, where $n, m, h$ and $w$ denote the number of images, number of channels, height and width, respectively.

(It seems like I'm not able to find something that I'm happy with, so I'll aim for being content ;) )

Regarding the "AXISNAME field `channel`", I agree as well. I was shifting this sentence around so many times and it just would not sound good, so this is what I settled with :grimacing: Would it be better like this?
The channels are identified via strings given in the `channel` [axis](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-axisname-field)  of the _NXdata_ class[^2].

Reviewer: I was also not in love with what I wrote, but I got bored.

I think the proposed mathematical notation is a bit problematic. The NXmx documentation uses a different notation (e.g. `nP`, `i`, `j`, `k`, `m`). Particularly, their use of `m` for channel is also a different sort of channel, they seem to use `m` for a chromatic channel of the incident beam (see NXmx documentation for `incident_wavelength`).

I think your updated text about the axis is an improvement.

Author: I see your point, that could be confusing with the $m$. How about this?

The dataset `/entry/data/data` has the four dimensions _number of images, number of channels, height, width_ (ordered from slowest to fastest changing direction).

Author: I changed the axis text in [78c44e8](https://github.com/soph-dec/documentation/pull/1/commits/78c44e868ca7569e7d0c0dcd49e3cdfdfb3137a5). The part with the dimensions/shape is still open.

Author: I realized I used notation like `nP, i, j` in the tables below already, so I added a suggestion to use this for the shape: [aad30da](https://github.com/soph-dec/documentation/pull/1/commits/aad30da4a31a418bd77bdce64f1850d1a4bdecde) and [b4d6551](https://github.com/soph-dec/documentation/pull/1/commits/b4d655156e7642e153952057ffeb3be817ecd0f3).
what do you think?

Reviewer: Looks good. I think using their notation is a reasonable and will probably avoid some confusion (even if i am not in love with their notation). `nC` for number of channels seems like a good fit.

Reviewer: There is still a remaining inconsistency though. Now both `i` and `j` and `height` and `width` are used (the "Example" section uses `nP` and `height` and `width`). It would be more consistent to use `i` and `j` everywhere, however, admittedly I think this would be less readable. Another possibility would be to use another notation for `i` and `j`, perhaps `nY` and `nX`? I don't know.

Author: Hmm, good catch. I think I would go with the `i` and `j` notation, even though I don't like that notation very much as well.

Author: As discussed, I used `i, j` everywhere now, see [6a5efd2](https://github.com/soph-dec/documentation/pull/1/commits/6a5efd24d4609f68248e711544b92c54d40b0a0f).

An _NXdetector_channel_ group could contain every field that is allowed for NXdetector and that makes sense to be given per channel.

[^2]: Currently, _AXISNAME_ is defined to be of type [_NX_NUMBER_](https://manual.nexusformat.org/nxdl-types.html#nx-number), but there is an [open PR](https://github.com/nexusformat/NIAC/issues/97) to allow [_NX_CHAR_](https://manual.nexusformat.org/nxdl-types.html#nx-char) as well.

#### Example

> For example, if `threshold_1`, `threshold_2` and `difference_mode` are enabled, we would get the following:

Reviewer: `difference_mode` -> `difference`

Author: Done in [5df342f](https://github.com/soph-dec/documentation/pull/1/commits/5df342f87f7c702e0adc7c8fe6081b5c9ecc2df1).

```
data: NXdata
  @signal = "data"
  @axes = ["image_number", "channel", ".", "."]
  @image_number_indices = 0
  @channel_indices = 1

>   image_number = [1, ..., number_of_images]

Reviewer: Our image IDs range from 0 to N-1, where N is the number of images.

 Also, instead of `image_number`, should we consider writing `image_id` since this is the spelling we use everywhere else?

Author: This array is actually `[0 + image_id_offset, ..., number_of_images - 1 + image_id_offset]`. Since `image_id_offset` defaults to `1`, what I wrote in the example is the default case.
In simplon v1 it was `image_nr_offset`, iirc this is why we first chose "image_number". I think it's a good idea to change it to "image_id" though, since that will match simplon v2.

Author: Regarding the renaming, it won't be part of the current bugfix release, so I would suggest to update the documentation when the change is in a release.

Reviewer: I totally forgot about `image_nr_offset`.

Thinking about this more, I think we should also consider (at least it should be discussed) solving this globally for SIMPLON V2. In particular, we should consider an `image_id_offset` that applies to all interfaces (not just Filewriter separately). Such a change may be relevant here in terms of what the default might be.

Author: The change for fw2 (image_number -> image_id) will be included in the next bugfix release, so I added the changes for the documentation in this [commit](https://github.com/soph-dec/documentation/pull/1/commits/353065b67577247ad220ebdae13780605164d14f).

Regarding the global `image_id_offset` for SIMPLON v2, I will open a ticket in Jira.

  channel = ["threshold_1", "threshold_2", "difference"]
  data = uint[height, width]

detector: NXdetector
  ...
  threshold_1_channel: NXdetector_channel
    threshold_energy = float
    flatfield = float[height, width]
    pixel_mask = uint[height, width]
  threshold_2_channel: NXdetector_channel
    threshold_energy = float
    flatfield = float[height, width]
    pixel_mask = uint[height, width]
  difference_channel: NXdetector_channel
    threshold_energy = float[2]
```

## List of fields

A file generated within V2 format may contain the fields and attributes listed below. For further details and descriptions please refer to the provided links to the NeXus definitions.

### / (root)

| Name | NeXus Type | Description |
|---|---|---|
| [@default](https://manual.nexusformat.org/classes/base_classes/NXroot.html?highlight=nxroot#nxroot-default-attribute) | NX_CHAR | Declares which NXentry group contains the data to be shown by default. |

> ### /entry

Reviewer: Link to https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-group

The top-level NeXus group contains all the data and associated information that comprise a single measurement.

| Name | NeXus Type | Description |
|---|---|---|

> | @NX_class | NX_CHAR | [`"NXentry"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-group)

Reviewer: Link directly to https://manual.nexusformat.org/classes/base_classes/NXentry.html#nxentry

| [definition](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-definition-field) | NX_CHAR | NeXus NXDL schema to which this file conforms, `"NXmx"`. |
| [end_time_estimated](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-end-time-estimated-field) | NX_DATE_TIME | Estimated time/date of the last data point collected in UTC, following ISO 8601 with suffix Z. Note that the time zone of the beamline is provided in `/entry/instrument/time_zone`. |
| [program_name](https://manual.nexusformat.org/classes/base_classes/NXentry.html#nxentry-program-name-field) | NX_CHAR | Name of the program used to generate this file, `"DECTRIS-DAQ"`.
| [program_name@version](https://manual.nexusformat.org/classes/base_classes/NXentry.html#nxentry-program-name-version-attribute) | NX_CHAR | Version number of the program used to generate this file. |
| [start_time](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-start-time-field) | NX_DATE_TIME | ISO 8601 time/date of the first data point collected in UTC. |

> ## /entry/data

Reviewer: Link to https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-data-group

Describes the plottable data and related dimension scales. Please also refer to [Multi-Channel Data](#multi-channel-data) for further details.

| Name | NeXus Type | Description |
|---|---|---|

> | @NX_class | NX_CHAR | [`"NXdata"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-data-group) |

Reviewer: Link directly to https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata

| [image_number](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-axisname-field) | NX_NUMBER[] | Dimension scale defining the axis `image_number` of the data, typically this is `[1, ..., nP]`. |
| [start_time](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-axisname-field) | NX_NUMBER[] | Dimension scale defining the axis `start_time` of the data. These are the relative start times for all images with absolute reference `/entry/start_time`. |
| [channel](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-axisname-field) | NX_CHAR[][^2] | Dimension scale defining the axis `channel` of the data. These are the enabled channels, for example `["threshold_1", "threshold_2", "difference"]`. |
| [data](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-data-data-field) | NX_NUMBER[nP,ch,i,j] | The image data.<br> **Note:** Files in V2 format always have only one `"/entry/data/data"` dataset, even in case `nimages_per_file` is greater than `0`. This is because it is a [Virtual Dataset](https://portal.hdfgroup.org/display/HDF5/Virtual+Dataset++-+VDS). |

> ## /entry/instrument

Reviewer: Link to https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-group

Collection of the components of the instrument or beamline.

| Name | NeXus Type | Description |
|---|---|---|

> | @NX_class | NX_CHAR | [`"NXinstrument"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-group) |

Reviewer: Link directly to https://manual.nexusformat.org/classes/base_classes/NXinstrument.html#nxinstrument

Author: Ah, I see, you would link the headings to NXmx and here link the base classes?

Reviewer: Yes, exactly.

Author: Done in [9a0aa15](https://github.com/soph-dec/documentation/pull/1/commits/9a0aa153fd707d28ac2deef3907139b566e42251)

| [name](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-name-field) | NX_CHAR | Name of the instrument. |
| [time_zone](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-time-zone-field) | NX_DATE_TIME | ISO 8601 time_zone offset from UTC. |

> ## /entry/instrument/beam

Reviewer: Link to https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-beam-group

Properties of the neutron or X-ray beam.

| Name | NeXus Type | Description |
|---|---|---|

> | @NX_class | NX_CHAR | [`"NXbeam"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-beam-group) |

Reviewer: Link directly to https://manual.nexusformat.org/classes/base_classes/NXbeam.html#nxbeam

| [incident_energy](https://manual.nexusformat.org/classes/base_classes/NXbeam.html?highlight=nxbeam#nxbeam-incident-energy-field) | NX_FLOAT | Energy carried by each particle of the beam on entering the beamline component. |
| incident_energy@units | NX_ENERGY | `"eV"` |
| [incident_wavelength](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-beam-incident-wavelength-field) | NX_FLOAT | Wavelength of the beam in the case of monochromatic beam. |
| incident_wavelength@units | NX_WAVELENGTH | `"angstrom"` |
| [total_flux](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-beam-total-flux-field) | NX_FLOAT | Flux incident on beam plane in photons per second. In other words, this is the flux integrated over the area. In the case of a beam that varies in flux shot-to-shot, this is an array of values, one for each recorded shot.|
| total_flux@units | NX_FREQUENCY | `"Hz"` |

> ## /entry/instrument/detector

Reviewer: Link to https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-group

| Name | NeXus Type | Description |
|---|---|---|

> | @NX_class | NX_CHAR | [`"NXdetector"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-group) |

Reviewer: Link directly to https://manual.nexusformat.org/classes/base_classes/NXdetector.html#nxdetector

Please also refer to the [_NXtransformations_](https://manual.nexusformat.org/classes/base_classes/NXtransformations.html#nxtransformations) documentation for further details on the representation of transformations and the `depens_on`-chain.

| Name | NeXus Type | Description |
|---|---|---|
| @NX_class | NX_CHAR | [`"NXtransformations"`](https://manual.nexusformat.org/classes/base_classes/NXtransformations.html#nxtransformations) |
| orientation | NX_FLOAT | Rotation angle of the detector. |
| orientation@depends_on | NX_CHAR | `"/entry/instrument/detector/geometry/translation"` |
| orientation@transformation_type | NX_CHAR | `"rotation"` |
| orientation@units | NX_CHAR | `"rad"` |
| orientation@vector | NX_FLOAT[3] | Unit vector describing the axis of the rotation. |
| translation | NX_FLOAT | Distance from the sample to the detector. |
| translation@depends_on | NX_CHAR | The transformation chain begins here, therefore this is `"."`. |
| translation@transformation_type | NX_CHAR | `"translation"` |
| translation@units | NX_CHAR | `"m"` |
| translation@vector | NX_FLOAT[3] | Unit vector describing the direction of the translation from the sample to the detector. |

> ## /entry/instrument/detector/module

Reviewer: Link to https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-detector-module-group

Geometry and logical description of a detector module. In our case, the full data array is represented as one single [_NXdetector_module_](https://manual.nexusformat.org/classes/base_classes/NXdetector_module.html#nxdetector-module).

| Name | NeXus Type | Description |
|---|---|---|

> | @NX_class | NX_CHAR | [`"NXdetector_module"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-detector-module-group) |

Reviewer: Link directly to https://manual.nexusformat.org/classes/base_classes/NXdetector_module.html#nxdetector-module

| [slow_pixel_direction@depends_on](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-detector-module-slow-pixel-direction-depends-on-attribute) | NX_CHAR | `"/entry/instrument/detector/transformations/translation"` |
| [slow_pixel_direction@offset](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-detector-module-slow-pixel-direction-offset-attribute) | NX_NUMBER[3] | A fixed offset applied before the transformation. |
| [slow_pixel_direction@transformation_type](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-detector-module-slow-pixel-direction-transformation-type-attribute) | NX_CHAR | `"translation"` |
| slow_pixel_direction@units | NX_LENGTH | `"m"` |
| [slow_pixel_direction@vector](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-detector-module-slow-pixel-direction-vector-attribute) | NX_NUMBER[3] | Unit vector describing the slowest varying pixel direction. |

> ## /entry/instrument/detector/transformations

Reviewer: Link to https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-transformations-group

| translation | NX_FLOAT | Distance from the sample to the detector. |
| translation@depends_on | NX_CHAR | `"."` |
| translation@offset | NX_NUMBER[3] | A fixed offset applied before the transformation. |
| translation@transformation_type | NX_CHAR | `"translation"` |
| translation@units | NX_LENGTH | `"m"` |
| translation@vector | NX_NUMBER[3] | Unit vector describing the direction of the translation from the sample to the detector. |
| two_theta | NX_FLOAT[nP] | Starting $2\theta$ value for each frame. |
| two_theta@offset | NX_NUMBER[3] | A fixed offset applied before the transformation. |
| two_theta@transformation_type | NX_CHAR | `"rotation"` |
| two_theta@units | NX_ANGLE | `"degree"` |
| two_theta@vector | NX_NUMBER[3] | Unit vector describing the axis of the rotation. |
| two_theta_end | NX_FLOAT[nP] | $2\theta$ end value for each frame. |
| two_theta_end@units | NX_ANGLE | `"degree"` |
| two_theta_increment_set | NX_FLOAT | The intended average range through which $2\theta$ moves during the exposure of a frame. |
| two_theta_increment_set@units | NX_ANGLE | `"degree"` |

> ## /entry/sample

Reviewer: Link to https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-sample-group

Group containing information on the sample.

| Name | NeXus Type | Description |
|---|---|---|

> | @NX_class | NX_CHAR | [`"NXsample"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-sample-group) |

Reviewer: Link to https://manual.nexusformat.org/classes/base_classes/NXsample.html#nxsample instead?

| [depends_on](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-sample-depends-on-field) | NX_CHAR | `"/entry/sample/transformations/omega"` |
| [name](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-sample-name-field) | NX_CHAR | Name of the sample. |

> ## /entry/sample/transformations

Reviewer: Link to https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-sample-transformations-group

Group containing sample goniometer and other related axes.

| Name | NeXus Type | Description |
|---|---|---|

> | @NX_class | NX_CHAR | [`"NXtransformations"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-sample-transformations-group) |

Reviewer: Link directly to https://manual.nexusformat.org/classes/base_classes/NXtransformations.html#nxtransformations

| kappa_increment_set@units | NX_ANGLE | `"degree"` |
| omega | NX_FLOAT[nP] | Starting $\omega$ value for each frame. |
| omega@depends_on | NX_CHAR | `"."` |
| omega@offset | NX_NUMBER[3] | A fixed offset applied before the transformation. |
| omega@transformation_type | NX_CHAR | `"rotation"` |
| omega@units | NX_ANGLE | `"degree"` |
| omega@vector | NX_NUMBER[3] | Unit vector describing the axis of the rotation. |
| omega_end | NX_FLOAT[nP] | $\omega$ end value for each frame. |
| omega_end@units | NX_ANGLE | `"degree"` |
| omega_increment_set | NX_FLOAT | The intended average range through which $\omega$ moves during the exposure of a frame. |
| omega_increment_set@units | NX_ANGLE | `"degree"` |
| phi | NX_FLOAT[nP] | Starting $\phi$ value for each frame. |
| phi@units | NX_ANGLE | `"degree"` |
| phi_end | NX_FLOAT[nP] | $\phi$ end value for each frame. |
| phi_end@units | NX_ANGLE | `"degree"` |
| phi_increment_set | NX_NUMBER | The intended average range through which $\phi$ moves during the exposure of a frame. |
| phi_increment_set@units | NX_ANGLE | `"degree"` |

> ## /entry/source

Reviewer: Link to https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-source-group

Group containing information about the neutron or x-ray storage ring/facility.

| Name | NeXus Type | Description |
|---|---|---|

> | @NX_class | NX_CHAR | [`"NXsource"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-source-group) |

Reviewer: Link directly to https://manual.nexusformat.org/classes/base_classes/NXsource.html#nxsource

# FileWriter V2

---
$\color{red}\text{\large{NOTE: FileWriter V2 is currently in a preview state and will change in the future.}}$

Feedback is very welcome!

---

FileWriter writes image data and metadata to [HDF5](https://www.hdfgroup.org/solutions/hdf5/) files. Currently, there are two different formats available. V1 is a legacy format that is loosely based on [version 1.4](https://github.com/nexusformat/definitions/releases/tag/NXmx-1.4) of the _NXmx application definition_ of the [NeXus] format. V2 follows the [_NXmx application definition_](https://manual.nexusformat.org/classes/applications/NXmx.html) of NeXus Version 2022.07, with the addition of multi-channel data support[^1].

| Version | Configuration | NeXus/NXmx Version | Multi-channel data support |
|----|----|----|----|
| V1 | SIMPLON | loosely based on [NXmx 1.4](https://github.com/nexusformat/definitions/releases/tag/NXmx-1.4) | no |
| V2 | SIMPLON | [NeXus 2022.07 NXmx](https://github.com/nexusformat/definitions/blob/v2022.07/applications/NXmx.nxdl.xml) | yes |

> [^1]: This is not part of the NeXus standard yet, a detailled PR will be opened when this data format has been approved by the community. The gerenal idea of this representation has already been discussed and approved by the NIAC though, see this [issue](https://github.com/nexusformat/definitions/issues/940).

Reviewer 2: typo: "general"

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

> [^1]: This is not part of the NeXus standard yet, a detailled PR will be opened when this data format has been approved by the community. The gerenal idea of this representation has already been discussed and approved by the NIAC though, see this [issue](https://github.com/nexusformat/definitions/issues/940).

Reviewer 3: typo: "detailed"

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

> [^1]: This is not part of the NeXus standard yet, a detailled PR will be opened when this data format has been approved by the community. The gerenal idea of this representation has already been discussed and approved by the NIAC though, see this [issue](https://github.com/nexusformat/definitions/issues/940).

Reviewer 4: ```suggestion
[^1]: This is not part of the NeXus standard yet, a detailled PR will be opened when this data format has been approved by the community. The general idea of this representation has already been discussed and approved by the NIAC though, see this [issue](https://github.com/nexusformat/definitions/issues/940).
```

> Therefore, a FileWriter V2 file is structured in the following groups:

Reviewer 2: structured into?

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

```
/
|
+-- entry
    |
    +-- data
    +-- instrument
    |   |
    |   +-- beam
    |   +-- detector
    |       |
    |       +-- threshold_1_channel*
    |       +-- threshold_2_channel*
    |       +-- difference_channel*
    |       +-- geometry
    |       +-- module
    |       +-- transformations
    |
    +-- sample
    |   |
    |   +-- transformations
    |   +-- beam
    |
    +-- source
```
\* this is not part of the NeXus standard yet, see [Multi-Channel Data](#multi-channel-data).

## FileWriter V2 Configuration

FileWriter V2 can be enabled and configured using the SIMPLON API.

### SIMPLON FileWriter Configuration Parameters

The configuration interface is accessible at `http://<ADDRESS_OF_DCU>/filewriter/api/1.8.0/config`.

| Parameter | JSON Type | Default | Description |
|-----------|------|---------|-------------|
| `mode` | string | `disabled` | Enabled state of the FileWriter, either `enabled` or `disabled`. |
| `compression_enabled` | bool | `true` | Optionally disable compression of image data. |
| `image_nr_start` | uint | `1` | First image number on the axis `/entry/image_number`. |

> | `name_pattern` | string | `series_$id` | Basename of the files. `$id` will be replaced by the series number, i.e. using the default, the generated files will have the names:<br> `series_<series_number>_master.h5`<br> `series_<series_number>_data_<file_number>.h5`|

Reviewer 4: ```suggestion
| `name_pattern` | string | `series_$id` | Base name of the files. `$id` will be replaced by the series number, i.e. using the default, the generated files will have the names:<br> `series_<series_number>_master.h5`<br> `series_<series_number>_data_<file_number>.h5`|
```

| `nimages_per_file` | uint | `1000` | Maximum number of images stored in each data file. If set to `0`, all images are stored directly in the master file and no data files are created. If set to a value greater than zero, the images are stored in multiple data files. |

> | `v2_preview_enabled` | bool | `false` | Enabled state of the FileWriter V2 preview. |

Reviewer 3: Should we additionally write that this will change? I mean the "preview" part makes it kinda obvious that it will change, but still, make this explicit. Might even be a different key altogether such as "format"

Author: I can add an extra warning, sure.

Author: ```suggestion
| `v2_preview_enabled` | bool | `false` | Enabled state of the FileWriter V2 preview. <br>**NOTE:** This key will be replaced when FileWriter V2 is released. It is likely that there will be a `format` key instead, in analogy to Stream V2. |
```

Author: How do you feel about that suggestion, @diego-gaemperle ?

Reviewer 3: I would rephrase the second sentence to:
"There will likely be..."
Otherwise sounds good to me

Author: Done in [2e366ce](https://github.com/soph-dec/documentation/pull/1/commits/2e366ceea5bace74784d3ceab0b02babd786a421)

<!-- SIMPLON V2 is currently not ready for filewriter

#### SIMPLON V2 FileWriter Configuration Parameters

The configuration interface is accessible at `http://<ADDRESS_OF_DCU>/api/v2/daq/config`.

| Parameter | JSON Type | Default | Description |
|-----------|------|---------|-------------|
| `enabled` | boolean | `false` | Enabled state of the FileWriter. |
| `image_id_offset` | uint | `1` | First image number on the axis `/entry/image_number`. |

> | `name_pattern` | string | `series_$id` | Basename of the files. `$id` will be replaced by the series number, i.e. using the default, the generated files will have the names:<br> `series_<series_number>_master.h5`<br> `series_<series_number>_data_<file_number>.h5`|

Reviewer 4: ```suggestion
| `name_pattern` | string | `series_$id` | Base name of the files. `$id` will be replaced by the series number, i.e. using the default, the generated files will have the names:<br> `series_<series_number>_master.h5`<br> `series_<series_number>_data_<file_number>.h5`|
```

> | `images_per_file` | uint | `1000` | Maximum number of images stored in each data file. If set to `0`, all images are stored directly in the master file and no data files are created. <br> If set to a value greater than zero, the images are stored in multiple data files, but are still accessible in the masterfile via the dataset "/entry/data/data", since this is a [Virtual Dataset](https://portal.hdfgroup.org/display/HDF5/Virtual+Dataset++-+VDS) in this case.|

Reviewer 4: ```suggestion
| `images_per_file` | uint | `1000` | Maximum number of images stored in each data file. If set to `0`, all images are stored directly in the master file and no data files are created. <br> If set to a value greater than zero, the images are stored in multiple data files, but are still accessible in the master file via the dataset "/entry/data/data", since this is a [Virtual Dataset](https://portal.hdfgroup.org/display/HDF5/Virtual+Dataset++-+VDS) in this case.|
```

| `v2_preview_enabled` | bool | `false` | Enabled state of the FileWriter V2 preview. |
-->

### Example

FileWriter V2 may be configured with cURL.

Using SIMPLON:
```sh
curl \
    -X PUT \
    -H "Content-Type: application/json" \
    -d '{"mode": {"value": "enabled"}, "v2_preview_enabled": {"value": true}}' \
    $ADDRESS_OF_DCU/filewriter/api/1.8.0/config
```

## Data Access

The files are stored locally on the DCU and have to be transferred to the user computer, this can also be done via cURL using the SIMPLON API.

```sh
# list available files
curl $ADDRESS_OF_DCU/filewriter/api/1.8.0/status/files

# file access
curl $ADDRESS_OF_DCU/data/<name_pattern>_master.h5

> curl $ADDRESS_OF_DCU/data/<name_pattern>_data_<file_number>.h5

Author: `-O` is missing here

Author: fixed in [f523944](https://github.com/soph-dec/documentation/pull/1/commits/f52394486c65b59cf5d9ca866f3236d073556a25)

```

For further details on available commands and status parameters please refer to the [SIMPLON API documentation](https://media.dectris.com/210607-DECTRIS-SIMPLON-API-Manual_EIGER2-chip-based_detectros.pdf).

## Multi-channel Data

One of the main new features of FileWriter V2 is the support of multi-channel data, i.e. each image consists of multiple channels, one per enabled threshold/difference mode. Therefore, the dataset `/entry/data/data` is four-dimensional, i.e. `[number_of_images, number_of_channels, width, height]`. The channels are identified via strings given in the [_AXISNAME_ field](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-axisname-field) `channel` of the _NXdata_ class[^2]. For every channel name listed in this field, there exists a subgroup of the [_NXdetector_]() group of type _NXdetector_channel_[^1] with the same name followed by the suffix `_channel`. This group contains channel-specific metadata, for example `threshold_energy`, `flatfield` or `pixel_mask`.

> An _NXdetector_channel_ group could contain every field that is allowed for NXdetector and that makes sense to be given per channel.

Reviewer 2: "A" group

Author: isn't it "an" because of the "NX" (which is pronounced "eneks"?)

Reviewer: I think "an" is correct.

Reviewer 4: this was interesting to find out:

> A and an are two different forms of the same word: the indefinite article a that is used before noun phrases. Use a when the noun or adjective that comes next begins with a consonant sound. Use an when the noun or adjective that comes next begins with a vowel sound. Remember that what matters is the pronunciation, not the spelling.

https://www.britannica.com/dictionary/eb/qa/how-do-you-know-whether-to-use-a-or-an

Reviewer 4: the spellchecker here, doesn't know that, and just suggests A because the word starts with N.

Author: I'll leave it as is then :)

[^2]: Currently, _AXISNAME_ is defined to be of type [_NX_NUMBER_](https://manual.nexusformat.org/nxdl-types.html#nx-number), but there is an [open PR](https://github.com/nexusformat/NIAC/issues/97) to allow [_NX_CHAR_](https://manual.nexusformat.org/nxdl-types.html#nx-char) as well.

#### Example

For example, if `threshold_1`, `threshold_2` and `difference_mode` are enabled, we would get the following:

```
data: NXdata
  @signal = "data"
  @axes = ["image_number", "channel", ".", "."]
  @image_number_indices = 0
  @channel_indices = 1
  image_number = [1, ..., number_of_images]
  channel = ["threshold_1", "threshold_2", "difference"]
  data = uint[height, width]

detector: NXdetector
  ...
  threshold_1_channel: NXdetector_channel
    threshold_energy = float
    flatfield = float[height, width]
    pixel_mask = uint[height, width]
  threshold_2_channel: NXdetector_channel
    threshold_energy = float
    flatfield = float[height, width]
    pixel_mask = uint[height, width]
  difference_channel: NXdetector_channel
    threshold_energy = float[2]
```

## List of fields

> A file generated with in V2 format may contain the fields and attributes listed below. For further details and descriptions please refer to the provided links to the NeXus definitions.

Reviewer 3: within

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

### / (root)

| Name | NeXus Type | Description |
|---|---|---|
| [@default](https://manual.nexusformat.org/classes/base_classes/NXroot.html?highlight=nxroot#nxroot-default-attribute) | NX_CHAR | Declares which NXentry group contains the data to be shown by default. |

### /entry

> The top-level NeXus group which contains all the data and associated information that comprise a single measurement.

Reviewer 3: Remove "which"

"The top-level NeXus group contains all the data and associated information that comprise a single measurement."

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

| Name | NeXus Type | Description |
|---|---|---|
| @NX_class | NX_CHAR | [`"NXentry"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-group)
| [@default](https://manual.nexusformat.org/classes/base_classes/NXentry.html#nxentry-default-attribute) | NX_CHAR | Declares which NXdata group contains the data to be shown by default. |
| [@version](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-version-attribute) | NX_CHAR | Describes the version of the NXmx definition used to write this data. |
| [definition](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-definition-field) | NX_CHAR | NeXus NXDL schema to which this file conforms, `"NXmx"`. |

> | [end_time_estimated](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-end-time-estimated-field) | NX_DATE_TIME | Estimated time/date of the last data point collected in UTC, following ISO 8601 with suffix Z. Note that the time zone of the beamline is provided in `/entry/instrument/time_zone`. |

Reviewer 4: ```suggestion
| [end_time_estimated](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-end-time-estimated-field) | NX_DATE_TIME | Estimated time/date of the last data point collected in UTC, following ISO 8601 with suffix Z. Note that the time zone of the beam-line is provided in `/entry/instrument/time_zone`. |
```

Reviewer: "beamline" is correct. See: https://en.wikipedia.org/wiki/Beamline

Author: Same here, won't change since it is correct

| [program_name](https://manual.nexusformat.org/classes/base_classes/NXentry.html#nxentry-program-name-field) | NX_CHAR | Name of the program used to generate this file, `"DECTRIS-DAQ"`.
| [program_name@version](https://manual.nexusformat.org/classes/base_classes/NXentry.html#nxentry-program-name-version-attribute) | NX_CHAR | Version number of the program used to generate this file. |
| [start_time](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-start-time-field) | NX_DATE_TIME | ISO 8601 time/date of the first data point collected in UTC. |

## /entry/data

> Describes the plottable data and related dimension scales. Please also refer to [Mutli-Channel Data](#multi-channel-data) for further details.

Reviewer 3: Multi-channel is written incorrectly in link name

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

> Describes the plottable data and related dimension scales. Please also refer to [Mutli-Channel Data](#multi-channel-data) for further details.

Reviewer 4: ```suggestion
Describes the plot table data and related dimension scales. Please also refer to [Mutli-Channel Data](#multi-channel-data) for further details.
```

Reviewer: "plottable" is an adjective and fine: https://en.wiktionary.org/wiki/plottable

Author: "plottable" is indeed what was meant here, therefore I will not change it.

Reviewer 4: oh, i thought it was an actual plot table, like a table with data to be plotted, my bad!
dunno why you would have a table with data to be plotted 😄

| Name | NeXus Type | Description |
|---|---|---|
| @NX_class | NX_CHAR | [`"NXdata"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-data-group) |
| [@axes](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-axes-attribute) | NX_CHAR[4] | String array that defines the independent data fields used in the default plot for all of the dimensions of the `/entry/data/data` field, `"image_number, channel, ".", "."`. |

> | [@image_number_indices](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-axisname-indices-attribute) | NX_INT | Indicates the dependency relationship of the `image_number` field with a dimensoin of the data, in this case `0`.

Reviewer 3: "dimension" instead of dimensoin

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

| [@start_time_indices](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-axisname-indices-attribute) | NX_INT | Indicates the dependency relationship of the `start_time` field with a dimensoin of the data, in this case `0`.

> | [@channel_indices](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-axisname-indices-attribute) | NX_INT | Indicates the dependency relationship of the `channel` field with a dimensoin of the data, in this case `1`.

Reviewer 3: also here "dimension" instead of dimensoin

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

| [start_time](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-axisname-field) | NX_NUMBER[] | Dimension scale defining the axis `start_time` of the data. These are the relative start times for all images with absolute reference `/entry/start_time`. |
| [channel](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-axisname-field) | NX_CHAR[][^2] | Dimension scale defining the axis `channel` of the data. These are the enabled channels, for example `["threshold_1", "threshold_2", "difference"]`. |
| [data](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-data-data-field) | NX_NUMBER[nP,ch,i,j] | The image data.<br> **Note:** Files in V2 format always have only one `"/entry/data/data"` dataset, even in case `nimages_per_file` is greater than `0`. This is because it is a [Virtual Dataset](https://portal.hdfgroup.org/display/HDF5/Virtual+Dataset++-+VDS). |

## /entry/instrument

Collection of the components of the instrument or beamline.

| Name | NeXus Type | Description |
|---|---|---|
| @NX_class | NX_CHAR | [`"NXinstrument"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-group) |

> | [name](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-name-field) | NX_CHAR | Name of instrument. |

Reviewer 3: Name of the instrument

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

| [time_zone](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-time-zone-field) | NX_DATE_TIME | ISO 8601 time_zone offset from UTC. |

## /entry/instrument/beam

Properties of the neutron or X-ray beam.

| Name | NeXus Type | Description |
|---|---|---|
| @NX_class | NX_CHAR | [`"NXbeam"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-beam-group) |

> | [incident_energy](https://manual.nexusformat.org/classes/base_classes/NXbeam.html?highlight=nxbeam#nxbeam-incident-energy-field) | NX_FLOAT | Energy carried by each particle of the beam on entering the beamline component. |

Reviewer 4: ```suggestion
| [incident_energy](https://manual.nexusformat.org/classes/base_classes/NXbeam.html?highlight=nxbeam#nxbeam-incident-energy-field) | NX_FLOAT | Energy carried by each particle of the beam on entering the beam-line component. |
```

Author: Since beamline is correct, I won't change this :)

| incident_energy@units | NX_ENERGY | `"eV"` |
| [incident_wavelength](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-beam-incident-wavelength-field) | NX_FLOAT | Wavelength of the beam in the case of monochromatic beam. |
| incident_wavelength@units | NX_WAVELENGTH | `"angstrom"` |

> | [total_flux](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-beam-total-flux-field) | NX_FLOAT | Flux incident on beam plane in photons per second. In other words this is the flux integrated over area. In the case of a beam that varies in flux shot-to-shot, this is an array of values, one for each recorded shot.|

Reviewer 3: "In other words, this is the flux integrated over the area"

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

| total_flux@units | NX_FREQUENCY | `"Hz"` |

## /entry/instrument/detector

| Name | NeXus Type | Description |
|---|---|---|
| @NX_class | NX_CHAR | [`"NXdetector"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-group) |
| [beam_center_x](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-beam-center-x-field) | NX_FLOAT | The x position of the direct beam on the detector. |
| beam_center_x@units | NX_LENGTH | `"pixel"` |
| [beam_center_y](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-beam-center-y-field) | NX_FLOAT | The y position of the direct beam on the detector. |
| beam_center_y@units | NX_LENGTH | `"pixel"` |
| [bit_depth_readout](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-bit-depth-readout-field) | NX_INT | Bit depth of the internal readout. |
| [count_time](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-count-time-field) | NX_FLOAT | Exposure time per image. |
| count_time@units | NX_TIME | `"s"` |

> | [countrate_correction_applied](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-countrate-correction-applied-field) | NX_BOOLEAN | Indicates whether countrate correction was applied. |

Reviewer 3: "count rate"

Author: I will change it to "count-rate" if that is okay with you (that is how it is in the nexus docs)

Reviewer: I think "count rate" is the more common spelling.

Author: OK, in that case I'll go with "count rate".

Author: done in [6ff24b6](https://github.com/soph-dec/documentation/pull/1/commits/6ff24b6ef02b0b9ab429b01d2a69bead2cf1ba3e)

| [depends_on](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-depends-on-field) | NX_CHAR | NeXus path to the detector positioner axis that most directly supports the detector. In the case of a single-module detector, the detector axis chain may start here. |
| [description](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-description-field) | NX_CHAR | Detector type and model information. |
| [detector_readout_time](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-detector-readout-time-field) | NX_FLOAT | Readout dead time between consecutive detector frames. |
| detector_readout_time@units | NX_TIME | `"s"` |
| [distance](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-distance-field) | NX_FLOAT | Distance from the sample to the beam center on the detector. |
| distance@units | NX_LENGTH | `"m"` |
| [flatfield_applied](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-flatfield-applied-field) | NX_BOOLEAN | Indicates whether a flatfield has been applied. |

> | [frame_time](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-frame-time-field) | NX_FLOAT | Time interval between start of image acquisitions. This defines the speed of data collection and is the inverse of the frame rate, the frequency of image acquisition. |

Reviewer 3: Time interval between the start ...

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

| frame_time@units | NX_TIME | `"s"` |

> | [number_of_cycles](https://manual.nexusformat.org/classes/base_classes/NXdetector.html#nxdetector-number-of-cycles-field) | NX_INT | When autosummation is enabled, each frame is constructed by summing multple subframes. `number_of_cycles` is the number of summed subframes. If autosummation is disabled, this field has the value `1`. |

Reviewer 3: I think it should be "auto-summation" (according to my auto-correct at least xD)
Also "multiple subframes"

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

| [saturation_value](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-saturation-value-field) | NX_INT | The value at which the detector goes into saturation. Data above this value is known to be invalid. |
| [sensor_material](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-sensor-material-field) | NX_CHAR | Material used for direct detection of X-rays in the sensor. |
| [sensor_thickness](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-sensor-thickness-field) | NX_FLOAT | Thickness of the sensor material. |
| sensor_thickness@units | NX_LENGTH | `"m"` |
| [serial_number](https://manual.nexusformat.org/classes/base_classes/NXdetector.html#nxdetector-serial-number-field) | NX_CHAR | Serial number of the detector. |
| [start_time](https://manual.nexusformat.org/classes/base_classes/NXdetector.html#nxdetector-start-time-field) | NX_FLOAT[nP] | Relative start time for each frame, with `/entry/start_time` as absolute reference. |

> | [virtual_pixel_interpolation_applied](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-virtual-pixel-interpolation-applied-field) | NX_BOOLEAN | Indicates wheter virtual pixel interpolation has been applied. |

Reviewer 3: whether

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

> | [x_pixel_size](https://manual.nexusformat.org/classes/base_classes/NXdetector.html#nxdetector-x-pixel-size-field) | NX_FLOAT | Size of a single pixel along x-axis of the detector. |

Reviewer 3: the x-axis

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

| x_pixel_size@units | NX_LENGTH | `"m"` |

> | [y_pixel_size](https://manual.nexusformat.org/classes/base_classes/NXdetector.html#nxdetector-y-pixel-size-field) | NX_FLOAT | Size of a single pixel along y-axis of the detector. |

Reviewer 3: the y-axis

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

| y_pixel_size@units | NX_LENGTH | `"m"` |

## /entry/instrument/detector/CHANNELNAME_channel

> For every channel enabled, there will be one _NXdetector_channel_ group containing metadata specific to that channel. For further details see [Mutli-Channel Data](#multi-channel-data).

Reviewer 3: Link name wrong:
"Multi-Channel ..."

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

| translation | NX_FLOAT | Distance from the sample to the detector. |
| translation@depends_on | NX_CHAR | The transformation chain begins here, therefore this is `"."`. |
| translation@transformation_type | NX_CHAR | `"translation"` |
| translation@units | NX_CHAR | `"m"` |
| translation@vector | NX_FLOAT[3] | Unit vector describing the direction of the translation from the sample to the detector. |

## /entry/instrument/detector/module

Geometry and logical description of a detector module. In our case, the full data array is represented as one single [_NXdetector_module_](https://manual.nexusformat.org/classes/base_classes/NXdetector_module.html#nxdetector-module).

| Name | NeXus Type | Description |
|---|---|---|
| @NX_class | NX_CHAR | [`"NXdetector_module"`](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-detector-module-group) |
| [data_origin](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-detector-module-data-origin-field) | NX_INT[2] | Indices of the data origin. The order of indices is slow to fast. |

> | [data_size](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-detector-module-data-size-field) | NX_INT[2] | Size of one channel image in pixels, the order of incies is slow to fast. |

Reviewer 3: indices

Author: fixed in [34890fd](https://github.com/soph-dec/documentation/pull/1/commits/34890fdc03430a5b054a735918d0593e38a8e157)

| slow_pixel_direction@units | NX_LENGTH | `"m"` |
| [slow_pixel_direction@vector](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-detector-module-slow-pixel-direction-vector-attribute) | NX_NUMBER[3] | Unit vector describing the slowest varying pixel direction. |

## /entry/instrument/detector/transformations

| Name | NeXus Type | Description |
|---|---|---|
|  |  |
| @NX_class | NX_CHAR | [`"NXtransformations"`](https://manual.nexusformat.org/classes/base_classes/NXtransformations.html) |
| translation | NX_FLOAT | Distance from the sample to the detector. |
| translation@depends_on | NX_CHAR | `"."` |
| translation@offset | NX_NUMBER[3] | A fixed offset applied before the transformation. |
| translation@transformation_type | NX_CHAR | `"translation"` |
| translation@units | NX_LENGTH | `"m"` |
| translation@vector | NX_NUMBER[3] | Unit vector describing the direction of the translation from the sample to the detector. |
| two_theta | NX_FLOAT[nP] | Starting $2\theta$ value for each frame. |

> | two_theta@depends_on | NX_CHAR | `"/entry/sample/transformations/translation"` |

Author: I think this is a bug. This should not point to the sample transformations group, but to the translation here in the detector transformations group. I will open a PR for wildhaus and when that is accepted, change these docs accordingly.

Reviewer: Do we know what `two_theta` means exactly? Is it documented somewhere? Otherwise, it's kind of hard for me to know what the best action should be. Maybe all these goniometer angles should be simply "user defined" and the `depends_on` attribute should be avoided altogether?

Author: What I found is this:

![image](https://user-images.githubusercontent.com/73243774/205010466-d6b08774-bdb4-457f-8597-14b8f7ab68e4.png)

Since "our" two_theta is in the detector-transformations group, I assume we are talking about the $\alpha$ ( $2\theta_{D}$ ), this would also match that we group it together with the distance to the sample.

I am not 100% sure though.

Author: screenshot is taken from this [paper](http://prism.mit.edu/xray/documents/Bruker%20GADDS%20Intro%20to%202D%20Diffraction.pdf)

Author: @diego-gaemperle also found this:
![two_theta](https://user-images.githubusercontent.com/73243774/205029645-1c729c73-239c-423d-9d30-62759eaa9fac.png)

After that, I actually think detector/transformations/translation needs to depend_on two_theta, since in this case (opposed to the detector/geometry case) we want $T_{R} \cdot T_{t}$, with notation as in [NXtransformations](https://manual.nexusformat.org/classes/base_classes/NXtransformations.html#nxtransformations).

Reviewer: The diagrams imply that detector translation would have to `depends_on` `two_theta`, as you noted.

NOTES: The first diagram uses different axis labels for the lab coordinate system than DECTRIS. Particularly, in the first diagram the sample starts at X_L and is rotated around Z_L. In the DECTRIS system, the detector face has a default normal of -Z_L. This is not an issue since their coordinate system is consistent with ours. Also, the first document mentions two-theta and two-theta-D. Our two-theta corresponds to their two-theta-D.

If this is indeed the universal meaning for "two theta", then I suppose it makes sense to make the detector translation depend_on `two_theta`. If we do this, I think we should document the exact meaning of `two_theta` like in the diagrams above (obviously using the DECTRIS axis labels). This should also be discussed briefly with the PMs.

Reviewer: As to the original bug: We don't write the dataset `"/entry/sample/transformations/translation"` right? If not, then obviously `two_theta` should not depend on it. We should remove the `two_theta@depends_on` attribute.

Author: We came to the same conclusion. We removed the `two_theta@depends_on` for now, but we will ask people who are actually using this what they would expect and what their interpretation is.

Author: I created a ticket for this in Jira, therefore I'll resolve this conversation here.

| [x_pixel_size](https://manual.nexusformat.org/classes/base_classes/NXdetector.html#nxdetector-x-pixel-size-field) | NX_FLOAT | Size of a single pixel along the x-axis of the detector. |
| x_pixel_size@units | NX_LENGTH | `"m"` |
| [y_pixel_size](https://manual.nexusformat.org/classes/base_classes/NXdetector.html#nxdetector-y-pixel-size-field) | NX_FLOAT | Size of a single pixel along the y-axis of the detector. |
| y_pixel_size@units | NX_LENGTH | `"m"` |

## /entry/instrument/detector/CHANNELNAME_channel

For every channel enabled, there will be one _NXdetector_channel_ group containing metadata specific to that channel. For further details see [Multi-Channel Data](#multi-channel-data).

| Name | NeXus Type | Description |
|---|---|---|
| @NX_class | NX_CHAR | `"NXdetector_channel"` |
| flatfield | NX_FLOAT[] | The flatfield applied to the channel data. |
| pixel_mask | NX_UINT[] | The pixel mask applied to the channel data. |
| threshold_energy | NX_FLOAT or NX_FLOAT[2] | The threshold energy for this channel. In case of difference mode, this is an array containing both the upper and lower threshold energies. |
| threshold_energy@units | NX_ENERGY | `"eV"` |

> ## [/entry/instrument/detector/geometry](https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-instrument-detector-transformations-group)

Reviewer: Why do we need both "geometry" and "transformations" groups? Should they be combined (e.g. only have "transformations")? Also, if we do want to keep goniometer parameters separate, we could also consider naming such a group "goniometer"?

Author: This is definitely subject to change. It should be discussed in a meeting Diego is organizing. But I agree, now that "geometry" is also of type "NXtransformations", we might group them together or name them differently.

| translation | NX_FLOAT | Distance from the sample to the detector. |
| translation@depends_on | NX_CHAR | `"."` |
| translation@offset | NX_NUMBER[3] | A fixed offset applied before the transformation. |
| translation@transformation_type | NX_CHAR | `"translation"` |
| translation@units | NX_LENGTH | `"m"` |
| translation@vector | NX_NUMBER[3] | Unit vector describing the direction of the translation from the sample to the detector. |
| two_theta | NX_FLOAT[nP] | Starting $2\theta$ value for each frame. |
| two_theta@offset | NX_NUMBER[3] | A fixed offset applied before the transformation. |
| two_theta@transformation_type | NX_CHAR | `"rotation"` |
| two_theta@units | NX_ANGLE | `"degree"` |
| two_theta@vector | NX_NUMBER[3] | Unit vector describing the axis of the rotation. |
| two_theta_end | NX_FLOAT[nP] | $2\theta$ end value for each frame. |
| two_theta_end@units | NX_ANGLE | `"degree"` |
| two_theta_increment_set | NX_FLOAT | The intended average range through which $2\theta$ moves during the exposure of a frame. |
| two_theta_increment_set@units | NX_ANGLE | `"degree"` |

> ## [/entry/sample]((https://manual.nexusformat.org/classes/applications/NXmx.html#nxmx-entry-sample-group))

Reviewer: Broken link (extra parentheses).

Author: Fixed in [32d64b2](https://github.com/soph-dec/documentation/pull/1/commits/32d64b26e8e47ee5a550d2c323a7966b39b922a5).

| V2 | SIMPLON | [NeXus 2022.07 NXmx](https://github.com/nexusformat/definitions/blob/v2022.07/applications/NXmx.nxdl.xml) | Yes |

[^1]: This is not part of the NeXus standard yet, a detailed PR will be opened when this data format has been approved by the community. The general idea of this representation has already been discussed and approved by the NIAC though, see this [issue](https://github.com/nexusformat/definitions/issues/940).

## FileWriter V2 Configuration

FileWriter V2 can be enabled and configured using the SIMPLON API.

### SIMPLON Configuration Parameters

The configuration interface is accessible at `http://<ADDRESS_OF_DCU>/filewriter/api/1.8.0/config`.

| Parameter | JSON Type | Default | Description |
|-----------|------|---------|-------------|
| `mode` | string | `"disabled"` | Enabled state of the FileWriter, either `"enabled"` or `"disabled"`. |
| `compression_enabled` | boolean | `true` | Optionally disable compression of image data. |
| `image_nr_start` | number | `1` | Unsigned integer defining the first image number on the axis `/entry/image_id`. |

> | `name_pattern` | string | `"series_$id"` | Base name of the files. `$id` will be replaced by the series number, i.e. using the default, the generated files will have the names:<br> `series_<series_number>_master.h5`<br> `series_<series_number>_data_<file_number>.h5`|

Reviewer: Replace uses of "series_number" with "series_id".

Author: done in [3180a11](https://github.com/soph-dec/documentation/pull/1/commits/3180a11e6d8ecbd3ffed1ada11e2a746b47eef15)

| `nimages_per_file` | number | `1000` | Unsigned integer defining the maximum number of images stored in each data file. If set to `0`, all images are stored directly in the master file and no data files are created. If set to a value greater than zero, the images are stored in multiple data files. |
| `v2_preview_enabled` | boolean | `false` | Enabled state of the FileWriter V2 preview. <br>**NOTE:** This key will be replaced when FileWriter V2 is released. There will likely be a `format` key instead, in analogy to Stream V2. |

<!-- SIMPLON V2 is currently not ready for filewriter

#### SIMPLON V2 Configuration Parameters

The configuration interface is accessible at `http://<ADDRESS_OF_DCU>/api/v2/daq/config`.

| Parameter | JSON Type | Default | Description |
|-----------|------|---------|-------------|
| `enabled` | boolean | `false` | Enabled state of the FileWriter. |
| `image_id_offset` | number | `1` | Unsigned integer defining the first image number on the axis `/entry/image_id`. |

> | `name_pattern` | string | `"series_$id"` | Base name of the files. `$id` will be replaced by the series number, i.e. using the default, the generated files will have the names:<br> `series_<series_number>_master.h5`<br> `series_<series_number>_data_<file_number>.h5`|

Reviewer: Replace uses of "series_number" with "series_id".

Author: done in [3180a11](https://github.com/soph-dec/documentation/pull/1/commits/3180a11e6d8ecbd3ffed1ada11e2a746b47eef15)

| `v2_preview_enabled` | bool | `false` | Enabled state of the FileWriter V2 preview. <br>**NOTE:** This key will be replaced when FileWriter V2 is released. There will likely be a `format` key instead, in analogy to Stream V2. |
-->

### Example

FileWriter V2 may be configured with cURL.

```sh
curl \
    -X PUT \
    -H "Content-Type: application/json" \
    -d '{"mode": {"value": "enabled"}, "v2_preview_enabled": {"value": true}}' \
    $ADDRESS_OF_DCU/filewriter/api/1.8.0/config
```

## Data Access

Files are stored locally (in memory) on the DCU and may be accessed over HTTP. Files are **not** stored persistently and all data will be lost on reboot.

```sh
# list available files
curl $ADDRESS_OF_DCU/filewriter/api/1.8.0/status/files

# file access
curl -O $ADDRESS_OF_DCU/data/<name_pattern>_master.h5
curl -O $ADDRESS_OF_DCU/data/<name_pattern>_data_<file_number>.h5
```

For further details on available commands and status parameters please refer to the [SIMPLON API documentation](https://media.dectris.com/210607-DECTRIS-SIMPLON-API-Manual_EIGER2-chip-based_detectros.pdf).

## Multi-Channel Data

> One of the main new features of FileWriter V2 is the support of multi-channel data, i.e. each image consists of multiple channels, one per enabled threshold/difference mode. The dataset `/entry/data/data` is four-dimensional with the shape `[nP, nC, i, j]`, the notation used is defined [here](#notation). The channels are identified via strings given in the `channel` [axis](https://manual.nexusformat.org/classes/base_classes/NXdata.html#nxdata-axisname-field) of the _NXdata_ class[^2]. For every channel name listed in this field, there exists a subgroup of the [_NXdetector_]() group of type _NXdetector_channel_[^1] with the same name followed by the suffix `_channel`. This group contains channel-specific metadata, for example `threshold_energy`, `flatfield` or `pixel_mask`.

Reviewer: You can drop the "the" before "shape".

Perhaps instead of "the notation used is defined here", the "Notation" subsection could be moved up and this remark could be avoided altogether? You probably considered this and couldn't find a good place for it? I don't have a good answer.

Author: Yes, it is as you suspected, I couldn't find a really good place for it.

Author: "the" removed in [92ac462](https://github.com/soph-dec/documentation/pull/1/commits/92ac462f59e9a78828e01acde45d262271abf443)
